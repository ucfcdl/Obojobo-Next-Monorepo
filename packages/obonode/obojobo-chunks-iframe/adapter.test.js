import IFrameAdapter from './adapter'
import setProp from 'obojobo-document-engine/src/scripts/common/util/set-prop'

describe('IFrame adapter', () => {
	let model

	class MockOboModel {
		constructor(attrs = {}) {
			this.modelState = {}
			this.content = attrs
		}

		setStateProp(propName, defaultValue, transformValueFn, allowedValues) {
			setProp(
				this.modelState,
				this.content,
				propName,
				defaultValue,
				transformValueFn,
				allowedValues
			)

			return true
		}
	}

	test('construct builds without attributes', () => {
		const model = new MockOboModel()
		IFrameAdapter.construct(model)
		expect(model.modelState).toMatchSnapshot()
	})

	test('adapter sets modelState.fit to "scale" by default, "scale" or "scroll" when specified, "scroll" if not specified (and type="webpage") and "scale" if not specified (and type="media")', () => {
		model = new MockOboModel({ fit: 'sCroll' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scroll')

		model = new MockOboModel({ fit: 'SCALE' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scale')

		model = new MockOboModel({ fit: 'other' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scale')

		model = new MockOboModel({ type: 'webpage' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scroll')

		model = new MockOboModel({ type: 'media' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scale')

		model = new MockOboModel({ type: 'webpage', fit: 'scroll' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scroll')

		model = new MockOboModel({ type: 'media', fit: 'scroll' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scroll')

		model = new MockOboModel({ type: 'webpage', fit: 'scale' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scale')

		model = new MockOboModel({ type: 'media', fit: 'scale' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scale')

		model = new MockOboModel({ type: 'webpage', fit: 'other' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scroll')

		model = new MockOboModel({ type: 'media', fit: 'other' })
		IFrameAdapter.construct(model)
		expect(model.modelState.fit).toBe('scale')
	})

	test('adapter sets modelState.controls to "reload" by default, "zoom,reload,new-window,border" if not specified (and type="webpage"), "reload" if not specified (and type="media") and lower-cases and trims values if specified', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.controls).toEqual(['reload'])

		model = new MockOboModel({ type: 'webpage' })
		IFrameAdapter.construct(model)
		expect(model.modelState.controls).toEqual(['zoom', 'reload', 'new-window', 'border'])

		model = new MockOboModel({ type: 'media' })
		IFrameAdapter.construct(model)
		expect(model.modelState.controls).toEqual(['reload'])

		model = new MockOboModel({ controls: '  ExAmPle , zoom' })
		IFrameAdapter.construct(model)
		expect(model.modelState.controls).toEqual(['example', 'zoom'])
	})

	test('adapter sets modelState.src to content.src (or null if not specified)', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.src).toBe(null)

		model = new MockOboModel({ src: 'mocked-src' })
		IFrameAdapter.construct(model)
		expect(model.modelState.src).toBe('mocked-src')
	})

	test('adapter sets modelState.width to a number if specified and null otherwise', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.width).toBe(null)

		model = new MockOboModel({ width: 'invalid' })
		IFrameAdapter.construct(model)
		expect(model.modelState.width).toBe(null)

		model = new MockOboModel({ width: 123 })
		IFrameAdapter.construct(model)
		expect(model.modelState.width).toBe(123)
	})

	test('adapter sets modelState.height to a number if specified and null otherwise', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.height).toBe(null)

		model = new MockOboModel({ height: 'invalid' })
		IFrameAdapter.construct(model)
		expect(model.modelState.height).toBe(null)

		model = new MockOboModel({ height: 123 })
		IFrameAdapter.construct(model)
		expect(model.modelState.height).toBe(123)
	})

	test('adapter sets modelState.initialZoom to a number if specified and 1 otherwise', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.initialZoom).toBe(1)

		model = new MockOboModel({ initialZoom: 'invalid' })
		IFrameAdapter.construct(model)
		expect(model.modelState.initialZoom).toBe(1)

		model = new MockOboModel({ initialZoom: 123 })
		IFrameAdapter.construct(model)
		expect(model.modelState.initialZoom).toBe(123)
	})

	test('adapter sets modelState.autoload to true if content.autoload === true, otherwise false', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.autoload).toBe(false)

		model = new MockOboModel({ autoload: false })
		IFrameAdapter.construct(model)
		expect(model.modelState.autoload).toBe(false)

		model = new MockOboModel({ autoload: true })
		IFrameAdapter.construct(model)
		expect(model.modelState.autoload).toBe(true)
	})

	test('adapter sets modelState.title to content.title (or null if not specified)', () => {
		model = new MockOboModel({})
		IFrameAdapter.construct(model)
		expect(model.modelState.title).toBe(null)

		model = new MockOboModel({ title: 'mocked-src' })
		IFrameAdapter.construct(model)
		expect(model.modelState.title).toBe('mocked-src')
	})

	test('can be cloned', () => {
		const a = new MockOboModel()
		const b = new MockOboModel()

		IFrameAdapter.construct(a)
		IFrameAdapter.clone(a, b)

		expect(a).not.toBe(b)
		expect(a.modelState).toEqual(b.modelState)
	})

	test('toJSON builds a JSON representation', () => {
		model = new MockOboModel({ src: 'mockSrc' })
		const json = { content: {} }

		IFrameAdapter.construct(model)
		IFrameAdapter.toJSON(model, json)
		expect(json).toMatchSnapshot()
	})

	test('toText creates a text representation', () => {
		model = new MockOboModel({ src: 'mockSrc' })

		IFrameAdapter.construct(model)
		expect(IFrameAdapter.toText(model)).toMatch('mockSrc')
	})
})
