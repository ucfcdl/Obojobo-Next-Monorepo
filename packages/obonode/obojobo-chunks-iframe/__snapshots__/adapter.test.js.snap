// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IFrame adapter construct builds without attributes 1`] = `
Object {
  "autoload": false,
  "border": false,
  "controls": Array [
    "reload",
  ],
  "fit": "scale",
  "height": null,
  "initialZoom": 1,
  "sizing": "fixed",
  "src": null,
  "title": null,
  "type": "media",
  "width": null,
}
`;

exports[`IFrame adapter toJSON builds a JSON representation 1`] = `
Object {
  "content": Object {
    "autoload": false,
    "border": false,
    "controls": Array [
      "reload",
    ],
    "fit": "scale",
    "height": null,
    "initialZoom": 1,
    "sizing": "fixed",
    "src": "mockSrc",
    "title": null,
    "type": "media",
    "width": null,
  },
}
`;
